drivers/platform/maxim/maxim_uart.c

```diff
@@ -0,0 +1,353 @@
+*******************************************************************************/
+
+/******************************************************************************/
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include <errno.h>
+#include <stdlib.h>
+#include "maxim_stdio.h"
+#include "mxc_sys.h"
+#include "uart.h"
+#include "maxim_uart.h"
+#include "no-os/uart.h"
+#include "no-os/irq.h"
+#include "no-os/util.h"
+
+/**
+* @brief Callback descriptor that contains a function to be called when an
+* interrupt occurs.
+*/
+static struct callback_desc *cb[2];
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+void UART0_IRQHandler()
+{
+	if(!cb[0])
+		return;
+
+	mxc_uart_regs_t *uart_regs = MXC_UART0;
+	uint8_t	n_int = 0;
+	volatile uint32_t reg_int = uart_regs->int_fl;
+
+	/** Clear all interrupt flags */
+	uart_regs->int_fl = 0x0;
+
+	while(reg_int) {
+		if((reg_int & 1) && uart_regs->int_en && BIT(n_int)) {
+			void *ctx = cb[0]->ctx;
+			void *config = cb[0]->config;
+			cb[0]->callback(ctx, n_int, config);
+		}
+		n_int++;
+		reg_int >>= 1;
+	}
+}
+
+void UART1_IRQHandler()
+{
+	if(!cb[1])
+		return;
+
+	mxc_uart_regs_t *uart_regs = MXC_UART1;
+	uint8_t	n_int = 0;
+	volatile uint32_t reg_int = uart_regs->int_fl;
+
+	/** Clear all interrupt flags */
+	uart_regs->int_fl = 0x0;
+
+	while(reg_int) {
+		if((reg_int & 1) && (uart_regs->int_en & BIT(n_int))) {
+			void *ctx = cb[1]->ctx;
+			void *config = cb[1]->config;
+			cb[1]->callback(ctx, n_int, config);
+		}
+		n_int++;
+		reg_int >>= 1;
+	}
+}
+
+/**
+ * @brief Read data from UART device. Blocking function.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return positive number of received bytes in case of success, negative error code otherwise.
+ */
+int32_t uart_read(struct uart_desc *desc, uint8_t *data, uint32_t bytes_number)
+{
+	if(!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	int32_t err = 0;
+	uint32_t bytes_read = 0;
+
+	int32_t len = bytes_number;
+	err = UART_Read(MXC_UART_GET_UART(desc->device_id), data, len, &bytes_read);
+
+	if(err < 0)
+		return -EIO;
+
+	return bytes_read;
+}
+
+/**
+ * @brief Write data to UART device. Blocking function.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t uart_write(struct uart_desc *desc, const uint8_t *data,
+		   uint32_t bytes_number)
+{
+	if(!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	uint32_t err = 0;
+
+	int32_t len = bytes_number;
```

Comment:
I assume you have a maximum length to transfer with UART_Write, if so, I would split the transfer in multiple transfer.

Reply:
UART_Write handles that. It works by adding each byte to the tx fifo, and also waiting (in a loop) if the fifo is full. So in this case I don't think I have to split into multiple transfers.